File: Sample/Engine/UI/Rect.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
import copy
from typing import ClassVar, Dict, List, Optional, Tuple, Union, TYPE_CHECKING
from .. import (
    TypeAdapter,
    Pair,
    Sprite,
    IntRect,
    Vector2i,
    Vector2f,
    Texture,
    RenderTexture,
    RectBase,
    Curve,
)
from ..Utils import Math, Render
from .Base import SpriteBase

if TYPE_CHECKING:
    from Engine import Vector2u, Image


SELECTION_RECT_OPACITY_CURVE_KEY = "UI/SelectionRectOpacity"
_FALLBACK_FADE_SPEED = 96.0
_FALLBACK_OPACITY_RANGE: Tuple[float, float] = (128.0, 255.0)


class Rect(SpriteBase):
    r"""Rectangle widget with a fading opacity effect and skin-based rendering.

    Renders a rectangular area using a window skin texture with support
    for opacity fading animation driven by a curve asset.
    """

    _opacityCurves: ClassVar[Dict[str, Curve]] = {}

    @TypeAdapter(rect=([tuple, list], IntRect, lambda pos, size: IntRect(Vector2i(*pos), Vector2i(*size))))
    def __init__(
        self,
        rect: Union[IntRect, Tuple[Pair[int], Pair[int]], List[List[int]]],
        windowSkin: Image,
        opacityCurveKey: str = SELECTION_RECT_OPACITY_CURVE_KEY,
    ) -> None:
        r"""\brief Construct a Rect widget with curve-driven opacity.

        - \param rect             Logical position and size of the rectangle
        - \param windowSkin       Texture used for rendering the rectangle skin
        - \param opacityCurveKey  Curve asset key under Data/Curves
        """
        from .. import Scale

        self._size = Math.ToVector2u(rect.size)
        size = Math.ToVector2u(Math.ToVector2f(rect.size) * Scale)
        self._canvas: RenderTexture = RenderTexture(size)
        self._windowSkin = windowSkin
        self._rectImpl = RectBase()
        self._initUI()
        SpriteBase.__init__(self, self._canvas.getTexture())
        self.setPosition(Math.ToVector2f(rect.position))
        self._opacityCurveKey = opacityCurveKey
        self._opacityTime = 0.0
        self._opacity = float(self.getColour().a)
        self._opacityMultiplier = 1.0
        self._fading = True

    def setOpacityMultiplier(self, multiplier: float) -> None:
        r"""\brief Set a multiplier applied to the animated opacity.

        - \param multiplier  Opacity multiplier clamped between 0 and 1
        """
        self._opacityMultiplier = max(0.0, min(1.0, multiplier))
        self._applyOpacity()

    def getSize(self) -> Vector2u:
        r"""\brief Get the rectangle size in logical UI units.

        - \return  Rectangle size in logical UI units
        """
        return self._size

    def update(self, deltaTime: float) -> None:
        r"""\brief Update the curve-driven opacity animation.

        - \param deltaTime  Time elapsed since last update, in seconds
        """
        curve = self._getOpacityCurve()
        if curve is not None and curve.keys:
            duration = self._getOpacityDuration(curve)
            if duration > 0.0:
                self._opacityTime = (self._opacityTime + deltaTime) % duration
                sampleTime = float(curve.keys[0]["time"]) + self._opacityTime
                self._opacity = curve.evaluate(sampleTime)
            else:
                self._opacity = curve.evaluate(float(curve.keys[0]["time"]))
        else:
            self._updateFallbackOpacity(deltaTime)
        self._applyOpacity()

    def _getOpacityCurve(self) -> Optional[Curve]:
        if self._opacityCurveKey in Rect._opacityCurves:
            return Rect._opacityCurves[self._opacityCurveKey]
        try:
            from Source import Data

            curve = Data.getCurve(self._opacityCurveKey)
        except KeyError:
            return None
        Rect._opacityCurves[self._opacityCurveKey] = curve
        return curve

    def _getOpacityDuration(self, curve: Curve) -> float:
        if len(curve.keys) < 2:
            return 0.0
        return float(curve.keys[-1]["time"]) - float(curve.keys[0]["time"])

    def _updateFallbackOpacity(self, deltaTime: float) -> None:
        opacity = self._opacity
        opacityMin, opacityMax = _FALLBACK_OPACITY_RANGE
        if self._fading:
            opacity = max(opacityMin, opacity - _FALLBACK_FADE_SPEED * deltaTime)
            if opacity == opacityMin:
                self._fading = False
        else:
            opacity = min(opacityMax, opacity + _FALLBACK_FADE_SPEED * deltaTime)
            if opacity == opacityMax:
                self._fading = True
        self._opacity = opacity

    def _applyOpacity(self) -> None:
        colour = copy.copy(self.getColour())
        colour.a = int(max(0.0, min(255.0, self._opacity * self._opacityMultiplier)))
        self.setColour(colour)

    def _presave(self, target: List[Texture], area: List[IntRect]) -> None:
        target.clear()
        for i, _ in enumerate(area):
            sub_texture = Texture(self._windowSkin, False, area[i])
            target.append(sub_texture)

    def _initUI(self) -> None:
        self._windowEdge = RenderTexture(self._canvas.getSize())
        self._windowEdgeSprite = Sprite(self._windowEdge.getTexture())
        self._windowBack = Texture(self._windowSkin, False, Math.ToIntRect(132, 68, 24, 24))
        self._windowBackSprite = Sprite(self._windowBack)
        canvasSize = self._canvas.getSize()
        self._windowBackSprite.setScale(Vector2f(canvasSize.x / 24.0, canvasSize.y / 24.0))
        self._windowHintSprites: List[Sprite] = []
        self._pauseHintSprite: Sprite = None
        self._cachedCorners: List[Texture] = []
        self._cachedEdges: List[Texture] = []
        self._presave(
            self._cachedCorners,
            [
                Math.ToIntRect(128, 64, 4, 4),
                Math.ToIntRect(156, 64, 4, 4),
                Math.ToIntRect(128, 92, 4, 4),
                Math.ToIntRect(156, 92, 4, 4),
            ],
        )
        self._presave(
            self._cachedEdges,
            [
                Math.ToIntRect(132, 64, 24, 4),
                Math.ToIntRect(132, 92, 24, 4),
                Math.ToIntRect(128, 68, 4, 24),
                Math.ToIntRect(156, 68, 4, 24),
            ],
        )
        for edge in self._cachedEdges:
            edge.setRepeated(True)
        self._rectImpl.render(
            self._canvas,
            self._windowEdge,
            self._windowEdgeSprite,
            self._windowBackSprite,
            self._cachedCorners,
            self._cachedEdges,
            Render.CanvasRenderStates(),
        )

File: Sample/Core/Engine/src/UI/Rect.cpp
#include <UI/Rect.hpp>

#include <Runtime/EngineState.hpp>
#include <Runtime/RuntimeValue.hpp>
#include <Utils/Render.hpp>

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <stdexcept>

namespace {

constexpr float FallbackFadeSpeed = 96.0f;
constexpr float FallbackOpacityMin = 128.0f;
constexpr float FallbackOpacityMax = 255.0f;

}  // namespace

const std::string Rect::SelectionRectOpacityCurveKey =
    "UI/SelectionRectOpacity";

std::unordered_map<std::string, std::shared_ptr<Curve>> Rect::opacityCurves_;

void Rect::clearOpacityCurveCache() noexcept {
    opacityCurves_.clear();
}

Rect::Rect(const sf::IntRect& rect, const sf::Image& windowSkin,
           std::optional<std::string> opacityCurveKey)
    : SpriteBase(placeholderTexture()),
      size_(static_cast<float>(std::max(0, rect.size.x)),
            static_cast<float>(std::max(0, rect.size.y))),
      canvas_(std::make_shared<sf::RenderTexture>(
          nonZeroRenderTextureSize(sf::Vector2u(
              static_cast<unsigned int>(std::round(size_.x * Scale)),
              static_cast<unsigned int>(std::round(size_.y * Scale)))))),
      windowSkin_(windowSkin),
      opacityCurveKey_(opacityCurveKey.value_or(SelectionRectOpacityCurveKey)) {
    setPremultipliedTexture(true);
    initialiseUi();
    bindCanvasTexture();
    setPosition(sf::Vector2f(rect.position));
    opacity_ = static_cast<float>(getColour().a);
}

void Rect::setOpacityMultiplier(float multiplier) {
    opacityMultiplier_ = std::clamp(multiplier, 0.0f, 1.0f);
    applyOpacity();
}

void Rect::resize(const sf::Vector2f& size) {
    const sf::Vector2f logicalSize{
        std::max(0.0f, size.x),
        std::max(0.0f, size.y),
    };
    const sf::Vector2u logicalTextureSize{
        static_cast<unsigned int>(std::round(logicalSize.x * Scale)),
        static_cast<unsigned int>(std::round(logicalSize.y * Scale)),
    };
    const sf::Vector2u backingTextureSize =
        nonZeroRenderTextureSize(logicalTextureSize);
    if (size_ == logicalSize && canvas_->getSize() == backingTextureSize) {
        return;
    }
    size_ = logicalSize;
    if (!canvas_->resize(backingTextureSize)) {
        throw std::runtime_error(
            "Failed to resize selection rectangle render texture");
    }
    initialiseUi();
    bindCanvasTexture();
}

void Rect::setWindowSkin(const sf::Image& windowSkin) {
    windowSkin_ = windowSkin;
    initialiseUi();
    bindCanvasTexture();
}

sf::Vector2f Rect::getSize() const {
    return size_;
}

void Rect::update(float deltaTime) {
    const std::shared_ptr<Curve> curve = opacityCurve();
    if (curve != nullptr && !curve->keys.empty()) {
        const float duration = curve->getDuration();
        if (duration > 0.0f) {
            opacityTime_ = std::fmod(opacityTime_ + deltaTime, duration);
            opacity_ = curve->evaluate(curve->keys.front().time + opacityTime_);
        } else {
            opacity_ = curve->evaluate(curve->keys.front().time);
        }
    } else {
        updateFallbackOpacity(deltaTime);
    }
    applyOpacity();
}

void Rect::refreshDisplayScale() {
    resize(size_);
    ControlBase::refreshDisplayScale();
}

std::shared_ptr<sf::Texture> Rect::placeholderTexture() {
    return std::make_shared<sf::Texture>();
}

std::shared_ptr<Curve> Rect::resolveOpacityCurve(const std::string& key) {
    const auto cached = opacityCurves_.find(key);
    if (cached != opacityCurves_.end()) {
        return cached->second;
    }
    const std::vector<RuntimeValue> resolved =
        resolveRuntime("curve", {RuntimeValue(key)});
    if (resolved.empty()) {
        return nullptr;
    }
    const RuntimeValue::Object* object =
        resolved.front().getIf<RuntimeValue::Object>();
    if (object == nullptr) {
        return nullptr;
    }
    const std::shared_ptr<Curve> curve =
        std::dynamic_pointer_cast<Curve>(*object);
    if (curve != nullptr) {
        opacityCurves_.emplace(key, curve);
    }
    return curve;
}

void Rect::bindCanvasTexture() {
    std::shared_ptr<sf::Texture> texture(
        canvas_, const_cast<sf::Texture*>(&canvas_->getTexture()));
    setTexture(std::move(texture), true);
    const sf::Vector2u textureSize{
        static_cast<unsigned int>(std::round(size_.x * Scale)),
        static_cast<unsigned int>(std::round(size_.y * Scale)),
    };
    setTextureRect(
        {{0, 0},
         {static_cast<int>(textureSize.x), static_cast<int>(textureSize.y)}});
}

void Rect::initialiseUi() {
    cachedCornerTextures_.clear();
    cachedEdgeTextures_.clear();
    windowEdge_ = std::make_unique<sf::RenderTexture>(canvas_->getSize());
    windowBack_ = textureFromArea({{132, 68}, {24, 24}});
    windowEdgeSprite_ = std::make_unique<sf::Sprite>(windowEdge_->getTexture());
    windowBackSprite_ = std::make_unique<sf::Sprite>(*windowBack_);
    const sf::Vector2u canvasSize = canvas_->getSize();
    windowBackSprite_->setScale({static_cast<float>(canvasSize.x) / 24.0f,
                                 static_cast<float>(canvasSize.y) / 24.0f});

    const std::vector<sf::IntRect> corners = {
        {{128, 64}, {4, 4}},
        {{156, 64}, {4, 4}},
        {{128, 92}, {4, 4}},
        {{156, 92}, {4, 4}},
    };
    const std::vector<sf::IntRect> edges = {
        {{132, 64}, {24, 4}},
        {{132, 92}, {24, 4}},
        {{128, 68}, {4, 24}},
        {{156, 68}, {4, 24}},
    };
    std::vector<sf::Texture*> cachedCorners;
    std::vector<sf::Texture*> cachedEdges;
    cachedCorners.reserve(corners.size());
    cachedEdges.reserve(edges.size());
    for (const sf::IntRect& area : corners) {
        cachedCornerTextures_.push_back(textureFromArea(area));
        cachedCorners.push_back(cachedCornerTextures_.back().get());
    }
    for (const sf::IntRect& area : edges) {
        cachedEdgeTextures_.push_back(textureFromArea(area));
        cachedEdgeTextures_.back()->setRepeated(true);
        cachedEdges.push_back(cachedEdgeTextures_.back().get());
    }
    rectImpl_.render(*canvas_, *windowEdge_, *windowEdgeSprite_,
                     *windowBackSprite_, cachedCorners, cachedEdges,
                     canvasRenderStates());
}

std::unique_ptr<sf::Texture> Rect::textureFromArea(
    const sf::IntRect& area) const {
    return std::make_unique<sf::Texture>(windowSkin_, false, area);
}

std::shared_ptr<Curve> Rect::opacityCurve() const {
    return resolveOpacityCurve(opacityCurveKey_);
}

void Rect::updateFallbackOpacity(float deltaTime) {
    if (fading_) {
        opacity_ = std::max(opacity_ - FallbackFadeSpeed * deltaTime,
                            FallbackOpacityMin);
        if (opacity_ == FallbackOpacityMin) {
            fading_ = false;
        }
    } else {
        opacity_ = std::min(opacity_ + FallbackFadeSpeed * deltaTime,
                            FallbackOpacityMax);
        if (opacity_ == FallbackOpacityMax) {
            fading_ = true;
        }
    }
}

void Rect::applyOpacity() {
    sf::Color colour = getColour();
    const float alpha =
        std::floor(std::clamp(opacity_ * opacityMultiplier_, 0.0f, 255.0f));
    colour.a = static_cast<std::uint8_t>(alpha);
    setColour(colour);
}
